### ISSUE TITLE
Retry metrics

### ISSUE THREAD

**Issue by Aloren** [spring-projects/spring-retry#119]
We are using `spring-retry` the following way:
```
@Retryable(
        interceptor = "persistenceServiceRetryPolicy"
)
public interface PersistenceService {
 // bunch of persistence methods
}
```

We would like to have metrics for the number of retries for specific methods and specific implementations:
```
FirstPersistenceService.save
FirstPersistenceService.get
SecondPersistenceService.save
SecondPersistenceService.get
```

 `Label` in `@Retryable` is not usable for us as you can see, because we have annotation specified on the interface, it will be the same for all methods. By default if `label` is not specified, it fallbacks to `invocation.getMethod().toGenericString()`, which is also unusable: it returns first base class in the class hierarchy and method string must be parsed (this code will be duplicated for all services that use spring-retry) before usage. Is it possible to add `MethodInvocation` to the context the same way as `RetryContext.NAME` is added?

**Comment by dsyer** [comment]
The `RetryContext` is supposed to be `Serializable` I think, so adding a `MethodInvocation` might not be a smart move. If you have a better proposal for a default label in this case it would be interesting.

**Comment by mariusneo** [comment]
@dsyer I came across the same issue as @Aloren and went ahead to rewrite the `org.springframework.retry.interceptor.RetryOperationsInterceptor` class and add the method invocation to the retry context as well.
```
// RetryOperationsInterceptor.java#82
context.setAttribute(RetryContext.NAME, label);
```
```
context.setAttribute("context.method.invocation", invocation);
```

which I then use on `org.springframework.retry.RetryListener#close` for reporting retry usage patterns for the retryable methods (e.g. : `class.method` was called `n` times).

As seen in the RetrySynchronizationManager, the RetryContext used in managing retryable calls is actually living in `ThreadLocal`.

I am a bit confused about you saying :

> The RetryContext is supposed to be Serializable I think

IMO it should be left up to the user to choose what to take from the method invocation for monitoring purposes (maybe in some obscure cases it may make sense to use the values of the method parameters  for monitoring purposes).

**Comment by mariusneo** [comment]
>Another (naive) possibility which comes to my mind would be to avoid poluting the RetryContext with the method invocation and change the RetryCallback to contain a method to retrieve the MethodInvocation - but this is a rather awkward solution. So I'd rather discard this idea.

After looking more carefully at 
`org.springframework.retry.RetryContext#NAME` and its usages as well as to
`org.springframework.retry.interceptor.StatefulRetryOperationsInterceptor.MethodInvocationRetryCallback` I think that it is not a bad idea to actually expose a getter for the `invocation` in the `MethodInvocationRetryCallback` class.

I am starting a pull request after doing the required changes and the tests in the source code.

**Comment by dsyer** [comment]
Closed in #172
